watcher/src/watcher.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) 2018-2020 MobileCoin Inc.
+
+//! Basic Watcher Node
+
+use crate::{error::WatcherError, watcher_db::WatcherDB};
+
+use mc_api::conversions::block_num_to_s3block_path;
+use mc_common::{
+    logger::{log, Logger},
+    HashMap,
+};
+use mc_ledger_db::Ledger;
+use mc_ledger_sync::{ArchiveBlockData, ReqwestTransactionsFetcher};
+use mc_transaction_core::BlockSignature;
+
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread,
+    time::Duration,
+};
+
+/// Watches multiple consensus validators and collects block signatures.
+pub struct Watcher {
+    transactions_fetcher: Arc<ReqwestTransactionsFetcher>,
+    watcher_db: WatcherDB,
+    logger: Logger,
+}
+
+impl Watcher {
+    /// Create a new Watcher.
+    pub fn new(
+        watcher_db: WatcherDB,
+        transactions_fetcher: ReqwestTransactionsFetcher,
+        logger: Logger,
+    ) -> Self {
+        Self {
+            transactions_fetcher: Arc::new(transactions_fetcher),
+            watcher_db,
+            logger,
+        }
+    }
+
+    /// The number of blocks in the watcher db.
+    pub fn num_blocks(&self) -> u64 {
+        self.watcher_db.num_blocks().unwrap()
+    }
+
+    /// Sync blocks and collect signatures.
+    pub fn sync_signatures(&self, start: u64, end: Option<u64>) -> Result<(), WatcherError> {
+        let mut block_index = std::cmp::max(start, self.watcher_db.num_blocks()?);
+        log::debug!(
+            self.logger,
+            "Now syncing signatures from {} to {:?}",
+            block_index,
+            end,
+        );
+        loop {
+            if let Some(max_blocks) = end {
+                if block_index > max_blocks {
+                    return Ok(());
+                }
+            }
+
+            // Construct URL for the block we are trying to fetch.
+            let filename = block_num_to_s3block_path(block_index)
+                .into_os_string()
+                .into_string()
+                .unwrap();
+            let mut archive_blocks: HashMap<String, ArchiveBlockData> = HashMap::default();
+            let mut signatures: Vec<BlockSignature> = Vec::new();
+            for src_url in self.transactions_fetcher.source_urls.iter() {
+                let url = src_url.join(&filename)?;
+
+                // Try and get the block.
+                log::debug!(
+                    self.logger,
+                    "Attempting to fetch block {} from {}",
+                    block_index,
+                    url
+                );
+                match self.transactions_fetcher.block_from_url(&url) {
```

Comment:
What happens if this fails for one of the URLs? Not all nodes would upload at the exact same time.
Additionally, for example, if one node is down for 24 hours, syncing would halt for all other nodes.

Reply:
Good point - since the watcher's signature store is based entirely from archive blocks in S3, I'll store a map of "last_synced_archive_block" per URL in the watcher DB. 

Re: consensus validator node being down, to walk through what that will look like in the signature data, any new blocks will not have signatures from that node, because that node will get those from catchup, and did not participate in consensus. 

It is possible that the node may go down before publishing signed blocks to S3, so in that case, having the "last_synced_archive_block" will catch those.

Re: the watcher's view of network block height when using WatcherSync, it uses the local ledger_db, which was synced according to ledger_sync, based on the `is_blocking_and_quorum` evaluation of block height, so it's a decent indicator of network block height, without having to duplicate the logic in ledger sync, or make any requests to nodes for block height.

One other case to plan for is if the user restarts the watcher with a new consensus validator, then it will need to include that new node with a `last_synced_archive_block` of 0, and sync its signatures from S3.

Reply:
This is addressed!